system/lib/gl.c

```diff
@@ -1559,178 +1597,291 @@ void* emscripten_GetProcAddress(const char *name_) {
+    else if (!strcmp(name, "glResumeTransformFeedback")) return emscripten_glResumeTransformFeedback;
+    else if (!strcmp(name, "glSamplerParameterf")) return emscripten_glSamplerParameterf;
+    else if (!strcmp(name, "glSamplerParameterfv")) return emscripten_glSamplerParameterfv;
+    else if (!strcmp(name, "glSamplerParameteri")) return emscripten_glSamplerParameteri;
+    else if (!strcmp(name, "glSamplerParameteriv")) return emscripten_glSamplerParameteriv;
+    else if (!strcmp(name, "glTexImage3D")) return emscripten_glTexImage3D;
+    else if (!strcmp(name, "glTexStorage2D")) return emscripten_glTexStorage2D;
+    else if (!strcmp(name, "glTexStorage3D")) return emscripten_glTexStorage3D;
+    else if (!strcmp(name, "glTexSubImage3D")) return emscripten_glTexSubImage3D;
+    else if (!strcmp(name, "glTransformFeedbackVaryings")) return emscripten_glTransformFeedbackVaryings;
+    else if (!strcmp(name, "glUniform1ui")) return emscripten_glUniform1ui;
+    else if (!strcmp(name, "glUniform1uiv")) return emscripten_glUniform1uiv;
+    else if (!strcmp(name, "glUniform2ui")) return emscripten_glUniform2ui;
+    else if (!strcmp(name, "glUniform2uiv")) return emscripten_glUniform2uiv;
+    else if (!strcmp(name, "glUniform3ui")) return emscripten_glUniform3ui;
+    else if (!strcmp(name, "glUniform3uiv")) return emscripten_glUniform3uiv;
+    else if (!strcmp(name, "glUniform4ui")) return emscripten_glUniform4ui;
+    else if (!strcmp(name, "glUniform4uiv")) return emscripten_glUniform4uiv;
+    else if (!strcmp(name, "glUniformBlockBinding")) return emscripten_glUniformBlockBinding;
+    else if (!strcmp(name, "glUniformMatrix2x3fv")) return emscripten_glUniformMatrix2x3fv;
+    else if (!strcmp(name, "glUniformMatrix2x4fv")) return emscripten_glUniformMatrix2x4fv;
+    else if (!strcmp(name, "glUniformMatrix3x2fv")) return emscripten_glUniformMatrix3x2fv;
+    else if (!strcmp(name, "glUniformMatrix3x4fv")) return emscripten_glUniformMatrix3x4fv;
+    else if (!strcmp(name, "glUniformMatrix4x2fv")) return emscripten_glUniformMatrix4x2fv;
+    else if (!strcmp(name, "glUniformMatrix4x3fv")) return emscripten_glUniformMatrix4x3fv;
+    else if (!strcmp(name, "glUnmapBuffer")) return emscripten_glUnmapBuffer;
+    else if (!strcmp(name, "glVertexAttribDivisor")) return emscripten_glVertexAttribDivisor;
+    else if (!strcmp(name, "glVertexAttribI4i")) return emscripten_glVertexAttribI4i;
+    else if (!strcmp(name, "glVertexAttribI4iv")) return emscripten_glVertexAttribI4iv;
+    else if (!strcmp(name, "glVertexAttribI4ui")) return emscripten_glVertexAttribI4ui;
+    else if (!strcmp(name, "glVertexAttribI4uiv")) return emscripten_glVertexAttribI4uiv;
+    else if (!strcmp(name, "glVertexAttribIPointer")) return emscripten_glVertexAttribIPointer;
+    else if (!strcmp(name, "glWaitSync")) return emscripten_glWaitSync;
+  }
+#endif
+  
+#if GL_ASSERTIONS
```

Comment:
This is C code that goes to an Emscripten system library at system library build time, not at the time when user code is finally linked to form an output .js program. Therefore unlike library_x.js files, C code files do not get `#if USE_WEBGL2` and `#if GL_ASSERTIONS` settings into them from `src/settings.js`, but these will always be undefined.

The library is created here: https://github.com/emscripten-core/emscripten/blob/ec8830df5f85d8283315788ff83e3638468e80aa/tools/system_libs.py#L368.

To get the #defines into gl.c, instead of building just one `libgl.c` system library, it should probably be building separate `libwebgl1.bc` and `libwebgl2.bc` libraries in https://github.com/emscripten-core/emscripten/blob/ec8830df5f85d8283315788ff83e3638468e80aa/tools/system_libs.py#L643 . That way this file could get compiled twice for different values of USE_WEBGL2. Unfortunately in order to get GL_ASSERTIONS in, this would have to be compiled four times for each of the four combinations of the flag, so may be better to migrate this to unconditionally call out to a JS library function that has the `#if GL_ASSERTIONS` bit - then the decision to print an error or not could be deferred to a JS library code that does have the `GL_ASSERTIONS` check in it.